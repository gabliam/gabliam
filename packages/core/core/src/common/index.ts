export * from './utils';
export * from './interfaces';
