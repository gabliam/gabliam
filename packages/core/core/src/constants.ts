/** Type in registry */
export const TYPE = {
  Service: 'ServiceType',
  Config: 'ConfigType',
  PreDestroy: 'PreDestroy',
};

export const METADATA_KEY = {
  bean: '_bean',
  onMissingBean: '_onMissingBean',
  init: '_init',
  beforeCreate: '_beforeCreate',
  value: '_value',
  register: '_register',
  scan: '_scan',
  config: '_config',
  service: '_service',
  plugin: '_plugin',
  injectContainer: '_inject_container',
  preDestroy: '_preDestroy',
};

/**
 * Inject container key in descriptor
 */
export const INJECT_CONTAINER_KEY = '$$container';

export const ORDER_CONFIG = {
  Core: 50,
  Plugin: 100,
  Config: 150,
};

export const ERRORS_MSGS = {
  DUPLICATED_DECORATOR: `Cannot apply decorator multiple times.`,
  DUPLICATED_CONFIG_DECORATOR: `Cannot apply @Config, @PluginConfig or @CorePlugin decorator multiple times.`,
  DUPLICATED_PLUGIN_DECORATOR: `Cannot apply @Plugin decorator multiple times.`,
  DUPLICATED_APPLICATION_DECORATOR: `Cannot apply @Application decorator multiple times.`,
};

export const APP_CONFIG = Symbol('APP_CONFIG');

export const CORE_CONFIG = Symbol('CORE_CONFIG');

export const VALUE_EXTRACTOR = Symbol('ValueExtractor');
