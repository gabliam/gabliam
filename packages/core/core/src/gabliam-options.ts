import { GabliamConfig } from './interfaces';
import * as _ from 'lodash';

const DEFAULT_CONFIG: GabliamConfig = {
  config: process.env.GABLIAM_CONFIG_PATH,
};

export const getGabliamConfig = (
  options?: Partial<GabliamConfig> | string
): GabliamConfig => {
  if (options === undefined) {
    return DEFAULT_CONFIG;
  } else {
    if (_.isString(options)) {
      return {
        ...DEFAULT_CONFIG,
        config: options,
      };
    } else {
      return {
        ...DEFAULT_CONFIG,
        ...options,
      };
    }
  }
};
