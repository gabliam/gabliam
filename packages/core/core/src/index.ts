import 'reflect-metadata';

export {
  decorate,
  inject,
  injectable,
  interfaces as inversifyInterfaces,
  multiInject,
  named,
  optional,
  postConstruct,
  tagged,
  targetName,
  unmanaged,
} from 'inversify';
export * from './common';
export {
  APP_CONFIG,
  CORE_CONFIG,
  INJECT_CONTAINER_KEY,
  VALUE_EXTRACTOR,
} from './constants';
export * from './container';
export * from './decorator';
export * from './errors';
export * from './gabliam';
export * from './gabliam-options';
export * from './gabliam-utils';
export * from './interfaces';
export * from './joi';
export { configResolver, FileLoader, Loader, Resolver } from './loaders';
export * from './metadatas';
export * from './reflection';
export * from './registry';
export * from './value-extractor';
