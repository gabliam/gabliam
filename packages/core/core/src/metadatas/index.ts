export * from './bean';
export { Config, PluginConfig } from './config';
export * from './init';
export * from './inject-container';
export * from './lifecycle';
export * from './plugin';
export * from './register';
export * from './scan';
export * from './service';
export * from './value';
