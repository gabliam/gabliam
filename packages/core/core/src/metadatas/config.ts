/* eslint-disable @typescript-eslint/no-redeclare */
import { injectable } from 'inversify';
import { Type } from '../common';
import { ERRORS_MSGS, METADATA_KEY, ORDER_CONFIG, TYPE } from '../constants';
import { makeDecorator } from '../decorator';
import { Register } from './register';

/**
 * Type of the `Config` decorator / constructor function.
 */
export interface ConfigDecorator {
  /**
   * Decorator that marks a class as an Gabliam config and provides configuration
   * metadata that determines how the config should be processed,
   * instantiated.
   *
   * This class is loaded by the framework in config phase and all beans are injected in the container
   *
   * order of loading: CoreConfig -> PluginConfig -> Config
   *
   * @usageNotes
   *
   * Here is an example of a class that define a config
   *
   * ```typescript
   *  class Gretter {
   *      constructor(private name:string){};
   *      greet() {
   *          return `Hello ${this.name} !`;
   *      }
   *  }
   *
   * @Config()
   * class SampleConfig {
   *      @Bean(Gretter)
   *      createGretter() {
   *          return new Gretter('David');
   *      }
   * }
   * ```
   */
  (order?: number): ClassDecorator;

  /**
   * see the `@Config` decorator.
   */
  new (order?: number): any;
}

/**
 * `Config` decorator and metadata.
 */
export interface Config {
  /**
   * Order of loading
   */
  order: number;
}

const configDecorator = (defaultOrder: number): ConfigDecorator =>
  makeDecorator(
    METADATA_KEY.config,
    (order = defaultOrder) => ({ order }),
    (cls: Type<any>, annotationInstance: Config) => {
      injectable()(cls);
      Register({
        type: TYPE.Config,
        options: { order: annotationInstance.order },
      })(cls);
    },
    true,
    ERRORS_MSGS.DUPLICATED_CONFIG_DECORATOR,
  );

export const CoreConfig: ConfigDecorator = configDecorator(ORDER_CONFIG.Core);

export const PluginConfig: ConfigDecorator = configDecorator(
  ORDER_CONFIG.Plugin,
);

export const Config: ConfigDecorator = configDecorator(ORDER_CONFIG.Config);
