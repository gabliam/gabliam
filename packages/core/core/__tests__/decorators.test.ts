// tslint:disable:one-line
// tslint:disable:no-unused-expression
import Joi from 'joi';
import {
  Bean,
  BeforeCreate,
  Config,
  Init,
  InjectContainer,
  OnMissingBean,
  Plugin,
  PluginConfig,
  PreDestroy,
  reflection,
  Register,
  Scan,
  Service,
  Value,
} from '../src';
import { TYPE } from '../src/constants';
import { CoreConfig } from '../src/metadatas/config';

describe('@init', () => {
  test('should add init metadata to a class when decorating a method with @init', () => {
    @Config()
    class TestBean {
      @Init()
      testMethod() {}

      @Init()
      test2Method() {}
    }

    expect(
      reflection.propMetadataOfDecorator(TestBean, Init),
    ).toMatchSnapshot();
  });
});

describe('@BeforeCreate', () => {
  test('should add BeforeCreate metadata to a class when decorating a method with @BeforeCreate', () => {
    @Config()
    class TestBean {
      @BeforeCreate()
      testMethod() {}

      @BeforeCreate()
      test2Method() {}
    }

    expect(
      reflection.propMetadataOfDecorator(TestBean, BeforeCreate),
    ).toMatchSnapshot();
  });
});

describe('@InjectContainer', () => {
  test('should add InjectContainer metadata to a class when decorating a method with @InjectContainer', () => {
    @InjectContainer()
    @Config()
    class TestBean {}

    expect(
      reflection.annotationsOfDecorator(TestBean, InjectContainer),
    ).toMatchSnapshot();
  });
});

describe('@Bean', () => {
  test('should add Bean metadata to a class when decorated with @Bean', () => {
    class TestBean {
      @Bean('test')
      testMethod() {}

      @Bean('test2')
      test2Method() {}
    }
    expect(
      reflection.propMetadataOfDecorator(TestBean, Bean),
    ).toMatchSnapshot();
  });

  test('should add Bean metadata to a class when decorated multiple times with @Bean', () => {
    class TestBean {
      @Bean('test')
      testMethod() {}

      @Bean('test2')
      @Bean('test3')
      test2Method() {}
    }

    expect(
      reflection.propMetadataOfDecorator(TestBean, Bean),
    ).toMatchSnapshot();
  });
}); // end describe @Bean

describe('@OnMissingBean', () => {
  test('should add OnMissingBean metadata to a class when decorated with @OnMissingBean', () => {
    class TestBean {
      @OnMissingBean('OnMissingBean')
      @Bean('test')
      testMethod() {}

      @OnMissingBean('OnMissingBean2')
      @Bean('test2')
      test2Method() {}
    }

    expect(
      reflection.propMetadataOfDecorator(TestBean, OnMissingBean),
    ).toMatchSnapshot();
  });

  test('should add OnMissingBean metadata to a class when decorated multiple times with @OnMissingBean', () => {
    class TestBean {
      @OnMissingBean('OnMissingBean')
      @Bean('test')
      testMethod() {}

      @OnMissingBean('OnMissingBean1')
      @OnMissingBean('OnMissingBean2')
      @Bean('test2')
      @Bean('test3')
      test2Method() {}
    }

    expect(
      reflection.propMetadataOfDecorator(TestBean, OnMissingBean),
    ).toMatchSnapshot();
  });
}); // end describe @Bean

describe('@Config', () => {
  test('should add Registry and config metadata to a class when decorated with @Config', () => {
    @Config()
    class TestBean {}

    expect(reflection.annotations(TestBean)).toMatchSnapshot();
  });

  test('should add Registry and config metadata to a class when decorated with @Config(100)', () => {
    @Config(100)
    class TestBean {}

    expect(reflection.annotations(TestBean)).toMatchSnapshot();
  });

  test('should fail when decorated multiple times with @Config', () => {
    expect(function () {
      @Config()
      @Config()
      class TestBean {}

      new TestBean();
    }).toThrowError();
  });
}); // end describe @Config

describe('@PluginConfig', () => {
  test('should add Registry and config metadata to a class when decorated with @PluginConfig', () => {
    @PluginConfig()
    class TestBean {}

    expect(reflection.annotations(TestBean)).toMatchSnapshot();
  });

  test('should add Registry and config metadata to a class when decorated with @PluginConfig(100)', () => {
    @PluginConfig(100)
    class TestBean {}

    expect(reflection.annotations(TestBean)).toMatchSnapshot();
  });

  test('should fail when decorated multiple times with @Config', () => {
    expect(function () {
      @PluginConfig()
      @PluginConfig()
      class TestBean {}

      new TestBean();
    }).toThrowError();
  });
}); // end describe @Config

describe('@CoreConfig', () => {
  test('should add Registry and config metadata to a class when decorated with @CoreConfig', () => {
    @CoreConfig()
    class TestBean {}

    expect(reflection.annotations(TestBean)).toMatchSnapshot();
  });

  test('should add Registry and config metadata to a class when decorated with @CoreConfig(100)', () => {
    @CoreConfig(100)
    class TestBean {}

    expect(reflection.annotations(TestBean)).toMatchSnapshot();
  });

  test('should fail when decorated multiple times with @CoreConfig', () => {
    expect(function () {
      @CoreConfig()
      @CoreConfig()
      class TestBean {}

      new TestBean();
    }).toThrowError();
  });
}); // end describe @Config

describe('@Register', () => {
  test('should add Registry metadata to a class when decorated with @Register', () => {
    @Register({ type: TYPE.Config, id: TestBean })
    class TestBean {}

    expect(
      reflection.annotationsOfDecorator(TestBean, Register),
    ).toMatchSnapshot();
  });
}); // end describe @Register

describe('@Scan', () => {
  test('should add Scan metadata to a class when decorated with @Scan()', () => {
    @Scan()
    class TestBean {}

    expect(reflection.annotationsOfDecorator(TestBean, Scan)).toMatchSnapshot();
  });

  test('should add Scan metadata to a class when decorated with @Scan(relativePath)', () => {
    @Scan('./fixtures/loader')
    class TestBean {}

    expect(reflection.annotationsOfDecorator(TestBean, Scan)).toMatchSnapshot();
  });

  test('should add Scan metadata to a class when decorated with @Scan(relativePath) 2', () => {
    @Scan('../src')
    class TestBean {}

    expect(reflection.annotationsOfDecorator(TestBean, Scan)).toMatchSnapshot();
  });

  test('should add Scan metadata to a class when decorated with @Scan(__dirname)', () => {
    @Scan(__dirname)
    class TestBean {}

    expect(reflection.annotationsOfDecorator(TestBean, Scan)).toMatchSnapshot();
  });

  test('should add Scan metadata to a class when decorated multiple times with @Scan', () => {
    @Scan(__dirname)
    @Scan(`${__dirname}/otherFolder`)
    class TestBean {}

    expect(reflection.annotationsOfDecorator(TestBean, Scan)).toMatchSnapshot();
  });
}); // end describe @Scan

describe('@Service', () => {
  test('should add Service metadata to a class when decorated with @Service', () => {
    @Service()
    class TestBean {}

    expect(reflection.annotations(TestBean)).toMatchSnapshot();
  });

  test(`should add Service metadata to a class when decorated with @Service('Test)`, () => {
    @Service('Test')
    class TestBean {}

    expect(reflection.annotations(TestBean)).toMatchSnapshot();
  });
}); // end describe @Service

describe('@Value', () => {
  describe('@Value(options: ValueOptions)', () => {
    test('should add Value metadata to a class when decorated with @Value(options: ValueOptions)', () => {
      class TestBean {
        @Value({ path: 'application.name' })
        name: string;
        @Value({ path: 'application.surname', validator: Joi.string() })
        surname: string;
        @Value({
          path: 'application.surname',
          validator: { schema: Joi.string() },
        })
        firstname: string;
        @Value({
          path: 'application.postalcode',
          validator: {
            schema: Joi.string().required(),
            customErrorMsg: 'Error',
          },
        })
        postalcode: string;
        @Value({
          path: 'application.address',
          validator: Joi.string().required(),
        })
        address: string;
      }

      expect(
        reflection.propMetadataOfDecorator(TestBean, Value),
      ).toMatchSnapshot();
    });
  }); // end @Value(options: ValueOptions)

  test('should fail whit bad value', () => {
    expect(function () {
      class TestBean {
        @Value(<any>{ lol: 'application.name' })
        name: string;
      }

      new TestBean();
    }).toThrowError();
  });
}); // end describe @Value

describe('@Plugin', () => {
  test('should add Plugin metadata to a class when decorated with @Plugin()', () => {
    @Plugin()
    class TestBean {}
    expect(
      reflection.annotationsOfDecorator(TestBean, Plugin),
    ).toMatchSnapshot();
  });

  test(`should add Plugin metadata to a class when decorated with @Plugin('TestPlugin')`, () => {
    @Plugin('TestPlugin')
    class TestBean {}
    expect(
      reflection.annotationsOfDecorator(TestBean, Plugin),
    ).toMatchSnapshot();
  });

  test(`should add Plugin metadata to a class when decorated with @Plugin({ dependencies: ['TestPlugin'] })`, () => {
    @Plugin({ dependencies: ['TestPlugin'] })
    class TestBean {}
    expect(
      reflection.annotationsOfDecorator(TestBean, Plugin),
    ).toMatchSnapshot();
  });

  test(`should add Plugin metadata to a class when decorated with @Plugin({ name: 'TestPlugin' })`, () => {
    @Plugin({ name: 'TestPlugin' })
    class TestBean {}
    expect(
      reflection.annotationsOfDecorator(TestBean, Plugin),
    ).toMatchSnapshot();
  });

  test(`should add Plugin metadata to a class when decorated with @Plugin({ name: 'TestPlugin', dependencies: ['TestPlugin2'] })`, () => {
    @Plugin({ name: 'TestPlugin', dependencies: ['TestPlugin2'] })
    class TestBean {}
    expect(
      reflection.annotationsOfDecorator(TestBean, Plugin),
    ).toMatchSnapshot();
  });

  test('should fail when decorated multiple times with @Plugin', () => {
    expect(function () {
      @Plugin()
      @Plugin()
      class TestBean {}

      new TestBean();
    }).toThrowError();
  });

  test('should fail whit bad value', () => {
    expect(function () {
      @Plugin(<any>{ lol: 'application.name' })
      class TestBean {}

      new TestBean();
    }).toThrowError();
  });
});

describe('@preDestroy', () => {
  test('should add preDestroy metadata to a class when decorated with @preDestroy', () => {
    class TestBean {
      @PreDestroy()
      preDestroy() {}

      @PreDestroy()
      preDestroy2() {}
    }
    expect(
      reflection.propMetadataOfDecorator(TestBean, PreDestroy),
    ).toMatchSnapshot();
  });
});
