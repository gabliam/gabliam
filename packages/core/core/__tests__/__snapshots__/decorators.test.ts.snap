// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@Bean should add Bean metadata to a class when decorated multiple times with @Bean 1`] = `
Object {
  "test2Method": Array [
    PropDecoratorFactory {
      "id": "test2",
    },
    PropDecoratorFactory {
      "id": "test3",
    },
  ],
  "testMethod": Array [
    PropDecoratorFactory {
      "id": "test",
    },
  ],
}
`;

exports[`@Bean should add Bean metadata to a class when decorated with @Bean 1`] = `
Object {
  "test2Method": Array [
    PropDecoratorFactory {
      "id": "test2",
    },
  ],
  "testMethod": Array [
    PropDecoratorFactory {
      "id": "test",
    },
  ],
}
`;

exports[`@BeforeCreate should add BeforeCreate metadata to a class when decorating a method with @BeforeCreate 1`] = `
Object {
  "test2Method": Array [
    PropDecoratorFactory {},
  ],
  "testMethod": Array [
    PropDecoratorFactory {},
  ],
}
`;

exports[`@Config should add Registry and config metadata to a class when decorated with @Config 1`] = `
Array [
  DecoratorFactory {
    "order": 150,
  },
  DecoratorFactory {
    "autobind": true,
    "options": Object {
      "order": 150,
    },
    "type": "ConfigType",
  },
]
`;

exports[`@Config should add Registry and config metadata to a class when decorated with @Config(100) 1`] = `
Array [
  DecoratorFactory {
    "order": 100,
  },
  DecoratorFactory {
    "autobind": true,
    "options": Object {
      "order": 100,
    },
    "type": "ConfigType",
  },
]
`;

exports[`@CoreConfig should add Registry and config metadata to a class when decorated with @CoreConfig 1`] = `
Array [
  DecoratorFactory {
    "order": 50,
  },
  DecoratorFactory {
    "autobind": true,
    "options": Object {
      "order": 50,
    },
    "type": "ConfigType",
  },
]
`;

exports[`@CoreConfig should add Registry and config metadata to a class when decorated with @CoreConfig(100) 1`] = `
Array [
  DecoratorFactory {
    "order": 100,
  },
  DecoratorFactory {
    "autobind": true,
    "options": Object {
      "order": 100,
    },
    "type": "ConfigType",
  },
]
`;

exports[`@InjectContainer should add InjectContainer metadata to a class when decorating a method with @InjectContainer 1`] = `
Array [
  DecoratorFactory {},
]
`;

exports[`@OnMissingBean should add OnMissingBean metadata to a class when decorated multiple times with @OnMissingBean 1`] = `
Object {
  "test2Method": Array [
    PropDecoratorFactory {
      "id": "OnMissingBean1",
    },
    PropDecoratorFactory {
      "id": "OnMissingBean2",
    },
  ],
  "testMethod": Array [
    PropDecoratorFactory {
      "id": "OnMissingBean",
    },
  ],
}
`;

exports[`@OnMissingBean should add OnMissingBean metadata to a class when decorated with @OnMissingBean 1`] = `
Object {
  "test2Method": Array [
    PropDecoratorFactory {
      "id": "OnMissingBean2",
    },
  ],
  "testMethod": Array [
    PropDecoratorFactory {
      "id": "OnMissingBean",
    },
  ],
}
`;

exports[`@Plugin should add Plugin metadata to a class when decorated with @Plugin('TestPlugin') 1`] = `
Array [
  DecoratorFactory {
    "beforeAll": false,
    "dependencies": Array [],
    "name": "TestPlugin",
  },
]
`;

exports[`@Plugin should add Plugin metadata to a class when decorated with @Plugin() 1`] = `
Array [
  DecoratorFactory {
    "beforeAll": false,
    "dependencies": Array [],
    "name": undefined,
  },
]
`;

exports[`@Plugin should add Plugin metadata to a class when decorated with @Plugin({ dependencies: ['TestPlugin'] }) 1`] = `
Array [
  DecoratorFactory {
    "beforeAll": false,
    "dependencies": Array [
      Object {
        "name": "TestPlugin",
        "order": "after",
      },
    ],
    "name": undefined,
  },
]
`;

exports[`@Plugin should add Plugin metadata to a class when decorated with @Plugin({ name: 'TestPlugin' }) 1`] = `
Array [
  DecoratorFactory {
    "beforeAll": false,
    "dependencies": Array [],
    "name": "TestPlugin",
  },
]
`;

exports[`@Plugin should add Plugin metadata to a class when decorated with @Plugin({ name: 'TestPlugin', dependencies: ['TestPlugin2'] }) 1`] = `
Array [
  DecoratorFactory {
    "beforeAll": false,
    "dependencies": Array [
      Object {
        "name": "TestPlugin2",
        "order": "after",
      },
    ],
    "name": "TestPlugin",
  },
]
`;

exports[`@PluginConfig should add Registry and config metadata to a class when decorated with @PluginConfig 1`] = `
Array [
  DecoratorFactory {
    "order": 100,
  },
  DecoratorFactory {
    "autobind": true,
    "options": Object {
      "order": 100,
    },
    "type": "ConfigType",
  },
]
`;

exports[`@PluginConfig should add Registry and config metadata to a class when decorated with @PluginConfig(100) 1`] = `
Array [
  DecoratorFactory {
    "order": 100,
  },
  DecoratorFactory {
    "autobind": true,
    "options": Object {
      "order": 100,
    },
    "type": "ConfigType",
  },
]
`;

exports[`@Register should add Registry metadata to a class when decorated with @Register 1`] = `
Array [
  DecoratorFactory {
    "autobind": true,
    "id": [Function],
    "type": "ConfigType",
  },
]
`;

exports[`@Scan should add Scan metadata to a class when decorated multiple times with @Scan 1`] = `
Array [
  DecoratorFactory {
    "path": "<PROJECT_ROOT>/packages/core/core/__tests__/otherFolder",
  },
  DecoratorFactory {
    "path": "<PROJECT_ROOT>/packages/core/core/__tests__",
  },
]
`;

exports[`@Scan should add Scan metadata to a class when decorated with @Scan() 1`] = `
Array [
  DecoratorFactory {
    "path": "<PROJECT_ROOT>/packages/core/core/__tests__",
  },
]
`;

exports[`@Scan should add Scan metadata to a class when decorated with @Scan(__dirname) 1`] = `
Array [
  DecoratorFactory {
    "path": "<PROJECT_ROOT>/packages/core/core/__tests__",
  },
]
`;

exports[`@Scan should add Scan metadata to a class when decorated with @Scan(relativePath) 1`] = `
Array [
  DecoratorFactory {
    "path": "<PROJECT_ROOT>/packages/core/core/__tests__/fixtures/loader",
  },
]
`;

exports[`@Scan should add Scan metadata to a class when decorated with @Scan(relativePath) 2 1`] = `
Array [
  DecoratorFactory {
    "path": "<PROJECT_ROOT>/packages/core/core/src",
  },
]
`;

exports[`@Service should add Service metadata to a class when decorated with @Service 1`] = `
Array [
  DecoratorFactory {
    "name": undefined,
  },
  DecoratorFactory {
    "autobind": true,
    "id": [Function],
    "type": "ServiceType",
  },
]
`;

exports[`@Service should add Service metadata to a class when decorated with @Service('Test) 1`] = `
Array [
  DecoratorFactory {
    "name": "Test",
  },
  DecoratorFactory {
    "autobind": true,
    "id": "Test",
    "type": "ServiceType",
  },
]
`;

exports[`@Value @Value(options: ValueOptions) should add Value metadata to a class when decorated with @Value(options: ValueOptions) 1`] = `
Object {
  "address": Array [
    PropDecoratorFactory {
      "path": "application.address",
      "validator": Object {
        "schema": Object {
          "$_root": Object {
            "ValidationError": [Function],
            "_types": Set {
              "alternatives",
              "any",
              "array",
              "boolean",
              "date",
              "function",
              "link",
              "number",
              "object",
              "string",
              "symbol",
              "binary",
            },
            "allow": [Function],
            "alt": [Function],
            "alternatives": [Function],
            "any": [Function],
            "array": [Function],
            "assert": [Function],
            "attempt": [Function],
            "binary": [Function],
            "bool": [Function],
            "boolean": [Function],
            "build": [Function],
            "cache": Object {
              "provision": [Function],
            },
            "checkPreferences": [Function],
            "compile": [Function],
            "custom": [Function],
            "date": [Function],
            "defaults": [Function],
            "disallow": [Function],
            "equal": [Function],
            "exist": [Function],
            "expression": [Function],
            "extend": [Function],
            "forbidden": [Function],
            "func": [Function],
            "function": [Function],
            "in": [Function],
            "invalid": [Function],
            "isError": [Function],
            "isExpression": [Function],
            "isRef": [Function],
            "isSchema": [Function],
            "link": [Function],
            "not": [Function],
            "number": [Function],
            "object": [Function],
            "only": [Function],
            "optional": [Function],
            "options": [Function],
            "override": Symbol(override),
            "preferences": [Function],
            "prefs": [Function],
            "ref": [Function],
            "required": [Function],
            "string": [Function],
            "strip": [Function],
            "symbol": [Function],
            "trace": [Function],
            "types": [Function],
            "untrace": [Function],
            "valid": [Function],
            "version": "17.6.0",
            "when": [Function],
            "x": [Function],
            Symbol(@hapi/lab/coverage/initialize): [Function],
          },
          "$_super": Object {},
          "$_temp": Object {
            "ruleset": false,
            "whens": Object {},
          },
          "$_terms": Object {
            "alterations": null,
            "examples": null,
            "externals": null,
            "metas": Array [],
            "notes": Array [],
            "replacements": null,
            "shared": null,
            "tags": Array [],
            "whens": null,
          },
          "_cache": null,
          "_flags": Object {
            "presence": "required",
          },
          "_ids": Object {
            "_byId": Map {},
            "_byKey": Map {},
            "_schemaChain": false,
          },
          "_invalids": null,
          "_preferences": null,
          "_refs": Object {
            "refs": Array [],
          },
          "_rules": Array [],
          "_singleRules": Map {},
          "_valids": null,
          "type": "string",
        },
        "throwError": true,
      },
    },
  ],
  "firstname": Array [
    PropDecoratorFactory {
      "path": "application.surname",
      "validator": Object {
        "schema": Object {
          "$_root": Object {
            "ValidationError": [Function],
            "_types": Set {
              "alternatives",
              "any",
              "array",
              "boolean",
              "date",
              "function",
              "link",
              "number",
              "object",
              "string",
              "symbol",
              "binary",
            },
            "allow": [Function],
            "alt": [Function],
            "alternatives": [Function],
            "any": [Function],
            "array": [Function],
            "assert": [Function],
            "attempt": [Function],
            "binary": [Function],
            "bool": [Function],
            "boolean": [Function],
            "build": [Function],
            "cache": Object {
              "provision": [Function],
            },
            "checkPreferences": [Function],
            "compile": [Function],
            "custom": [Function],
            "date": [Function],
            "defaults": [Function],
            "disallow": [Function],
            "equal": [Function],
            "exist": [Function],
            "expression": [Function],
            "extend": [Function],
            "forbidden": [Function],
            "func": [Function],
            "function": [Function],
            "in": [Function],
            "invalid": [Function],
            "isError": [Function],
            "isExpression": [Function],
            "isRef": [Function],
            "isSchema": [Function],
            "link": [Function],
            "not": [Function],
            "number": [Function],
            "object": [Function],
            "only": [Function],
            "optional": [Function],
            "options": [Function],
            "override": Symbol(override),
            "preferences": [Function],
            "prefs": [Function],
            "ref": [Function],
            "required": [Function],
            "string": [Function],
            "strip": [Function],
            "symbol": [Function],
            "trace": [Function],
            "types": [Function],
            "untrace": [Function],
            "valid": [Function],
            "version": "17.6.0",
            "when": [Function],
            "x": [Function],
            Symbol(@hapi/lab/coverage/initialize): [Function],
          },
          "$_super": Object {},
          "$_temp": Object {
            "ruleset": null,
            "whens": Object {},
          },
          "$_terms": Object {
            "alterations": null,
            "examples": null,
            "externals": null,
            "metas": Array [],
            "notes": Array [],
            "replacements": null,
            "shared": null,
            "tags": Array [],
            "whens": null,
          },
          "_cache": null,
          "_flags": Object {},
          "_ids": Object {
            "_byId": Map {},
            "_byKey": Map {},
            "_schemaChain": false,
          },
          "_invalids": null,
          "_preferences": null,
          "_refs": Object {
            "refs": Array [],
          },
          "_rules": Array [],
          "_singleRules": Map {},
          "_valids": null,
          "type": "string",
        },
        "throwError": true,
      },
    },
  ],
  "name": Array [
    PropDecoratorFactory {
      "path": "application.name",
      "validator": null,
    },
  ],
  "postalcode": Array [
    PropDecoratorFactory {
      "path": "application.postalcode",
      "validator": Object {
        "customErrorMsg": "Error",
        "schema": Object {
          "$_root": Object {
            "ValidationError": [Function],
            "_types": Set {
              "alternatives",
              "any",
              "array",
              "boolean",
              "date",
              "function",
              "link",
              "number",
              "object",
              "string",
              "symbol",
              "binary",
            },
            "allow": [Function],
            "alt": [Function],
            "alternatives": [Function],
            "any": [Function],
            "array": [Function],
            "assert": [Function],
            "attempt": [Function],
            "binary": [Function],
            "bool": [Function],
            "boolean": [Function],
            "build": [Function],
            "cache": Object {
              "provision": [Function],
            },
            "checkPreferences": [Function],
            "compile": [Function],
            "custom": [Function],
            "date": [Function],
            "defaults": [Function],
            "disallow": [Function],
            "equal": [Function],
            "exist": [Function],
            "expression": [Function],
            "extend": [Function],
            "forbidden": [Function],
            "func": [Function],
            "function": [Function],
            "in": [Function],
            "invalid": [Function],
            "isError": [Function],
            "isExpression": [Function],
            "isRef": [Function],
            "isSchema": [Function],
            "link": [Function],
            "not": [Function],
            "number": [Function],
            "object": [Function],
            "only": [Function],
            "optional": [Function],
            "options": [Function],
            "override": Symbol(override),
            "preferences": [Function],
            "prefs": [Function],
            "ref": [Function],
            "required": [Function],
            "string": [Function],
            "strip": [Function],
            "symbol": [Function],
            "trace": [Function],
            "types": [Function],
            "untrace": [Function],
            "valid": [Function],
            "version": "17.6.0",
            "when": [Function],
            "x": [Function],
            Symbol(@hapi/lab/coverage/initialize): [Function],
          },
          "$_super": Object {},
          "$_temp": Object {
            "ruleset": false,
            "whens": Object {},
          },
          "$_terms": Object {
            "alterations": null,
            "examples": null,
            "externals": null,
            "metas": Array [],
            "notes": Array [],
            "replacements": null,
            "shared": null,
            "tags": Array [],
            "whens": null,
          },
          "_cache": null,
          "_flags": Object {
            "presence": "required",
          },
          "_ids": Object {
            "_byId": Map {},
            "_byKey": Map {},
            "_schemaChain": false,
          },
          "_invalids": null,
          "_preferences": null,
          "_refs": Object {
            "refs": Array [],
          },
          "_rules": Array [],
          "_singleRules": Map {},
          "_valids": null,
          "type": "string",
        },
        "throwError": true,
      },
    },
  ],
  "surname": Array [
    PropDecoratorFactory {
      "path": "application.surname",
      "validator": Object {
        "schema": Object {
          "$_root": Object {
            "ValidationError": [Function],
            "_types": Set {
              "alternatives",
              "any",
              "array",
              "boolean",
              "date",
              "function",
              "link",
              "number",
              "object",
              "string",
              "symbol",
              "binary",
            },
            "allow": [Function],
            "alt": [Function],
            "alternatives": [Function],
            "any": [Function],
            "array": [Function],
            "assert": [Function],
            "attempt": [Function],
            "binary": [Function],
            "bool": [Function],
            "boolean": [Function],
            "build": [Function],
            "cache": Object {
              "provision": [Function],
            },
            "checkPreferences": [Function],
            "compile": [Function],
            "custom": [Function],
            "date": [Function],
            "defaults": [Function],
            "disallow": [Function],
            "equal": [Function],
            "exist": [Function],
            "expression": [Function],
            "extend": [Function],
            "forbidden": [Function],
            "func": [Function],
            "function": [Function],
            "in": [Function],
            "invalid": [Function],
            "isError": [Function],
            "isExpression": [Function],
            "isRef": [Function],
            "isSchema": [Function],
            "link": [Function],
            "not": [Function],
            "number": [Function],
            "object": [Function],
            "only": [Function],
            "optional": [Function],
            "options": [Function],
            "override": Symbol(override),
            "preferences": [Function],
            "prefs": [Function],
            "ref": [Function],
            "required": [Function],
            "string": [Function],
            "strip": [Function],
            "symbol": [Function],
            "trace": [Function],
            "types": [Function],
            "untrace": [Function],
            "valid": [Function],
            "version": "17.6.0",
            "when": [Function],
            "x": [Function],
            Symbol(@hapi/lab/coverage/initialize): [Function],
          },
          "$_super": Object {},
          "$_temp": Object {
            "ruleset": null,
            "whens": Object {},
          },
          "$_terms": Object {
            "alterations": null,
            "examples": null,
            "externals": null,
            "metas": Array [],
            "notes": Array [],
            "replacements": null,
            "shared": null,
            "tags": Array [],
            "whens": null,
          },
          "_cache": null,
          "_flags": Object {},
          "_ids": Object {
            "_byId": Map {},
            "_byKey": Map {},
            "_schemaChain": false,
          },
          "_invalids": null,
          "_preferences": null,
          "_refs": Object {
            "refs": Array [],
          },
          "_rules": Array [],
          "_singleRules": Map {},
          "_valids": null,
          "type": "string",
        },
        "throwError": true,
      },
    },
  ],
}
`;

exports[`@init should add init metadata to a class when decorating a method with @init 1`] = `
Object {
  "test2Method": Array [
    PropDecoratorFactory {},
  ],
  "testMethod": Array [
    PropDecoratorFactory {},
  ],
}
`;

exports[`@preDestroy should add preDestroy metadata to a class when decorated with @preDestroy 1`] = `
Object {
  "preDestroy": Array [
    PropDecoratorFactory {},
  ],
  "preDestroy2": Array [
    PropDecoratorFactory {},
  ],
}
`;
